### DISCUSSION TITLE
API Gateway with RxProxyHttpClient or Spring Cloud Gateway.

### DISCUSSION THREAD

**Discussion by rafael-coelho-amaro** [micronaut-projects/micronaut-core#5311] [Q&A]
Hi everyone, I still having some issues with Micronaut RxProxyHttpClient. The point now is that this client just not redirect the API request, but change something into it (the request). So when I am doing a HTTPS request the third party API return the HTTP ERROR 421 (Misdirected Request).
For comparison, when using Spring Cloud Gateway, the same request works as wished.
So, I really need someone helps to keep forward with Micronaut, I doin't wanna to switch the application to Spring.
Then I need to know if it is possible do proxy a API request without any changes on the request. Or if it is possible to integrate Spring Cloud Gateway with Micronaut (so I don't need to switch my hole application to Spring).

Please, anyone could help?

**Comment by rafael-coelho-amaro** [comment]
So, I am create a API Gateway using Micronaut.  I am using  **micronaut-http-client** and **micronaut-http-server-netty** for that. Then I am using  the client ** @Client("\${gateway-conf.hybris-url}", httpVersion = HttpVersion.HTTP_1_1)** for inject **RxProxyHttpClient**.
The point is, when I call a request using simple HTTP request, it works OK, but HTTPS validates on security layer if the requests hade any changes. If it has any changes it returns the HTTP error 421.
```
I have, for example, 2 endpoints that I will redirect.
http://localhost:8080/hybris/v2/amaro-br/users/anonymous/carts/  - micronaut proxy
https://localhost:8080/hybris/v2/amaro-br/users/anonymous/carts/  - micronaut proxy with https
```
```
and both of them must be proxied to.
http://localhost:8083/hybris/v2/amaro-br/users/anonymous/carts/  - third party API
https://localhost:8083/hybris/v2/amaro-br/users/anonymous/carts/  - third party API
```
```
So it must redirect like that.
http://localhost:8080/hybris/v2/amaro-br/users/anonymous/carts/ to http://localhost:8080/hybris/v2/amaro-br/users/anonymous/carts/ 
https://localhost:8080/hybris/v2/amaro-br/users/anonymous/carts/  to https://localhost:8083/hybris/v2/amaro-br/users/anonymous/carts/
```

When the call is made using simple HTTP, it works well, if it is call using HTTPS I get error 421.
So check if it is possible to execute those proxies, I did same thing using Spring cloud, and works perfect.

My micronaur code for HTTPS proxy:
```
package com.gateways.controllers

import io.micronaut.http.HttpRequest
import io.micronaut.http.HttpVersion
import io.micronaut.http.MutableHttpResponse
import io.micronaut.http.annotation.Filter
import io.micronaut.http.client.RxProxyHttpClient
import io.micronaut.http.client.annotation.Client
import io.micronaut.http.filter.OncePerRequestHttpServerFilter
import io.micronaut.http.filter.ServerFilterChain
import org.reactivestreams.Publisher


@Filter("/hybris/**")
class ProxyFilter (@Client("https://localhost:8083",
                            httpVersion = HttpVersion.HTTP_1_1)
                            val client: RxProxyHttpClient) :
        OncePerRequestHttpServerFilter() {

    override fun doFilterOnce(request: HttpRequest<*>?,
                              chain: ServerFilterChain?):
            Publisher<MutableHttpResponse<*>> {
        return client.proxy(request!!.mutate())
    }
}
```

Sample using Spring Cloud Gateway (using **spring-cloud-starter-gateway**, that I just need to configure the filters and redirects):
```
spring:
  cloud:
    gateway:
      routes:
        - id: hybris_proxy
          uri: https://localhost:8083
          predicates:
            - Path=/hybris/**
```

**Reply by graemerocher** [reply]
Can you add a logger for `io.micronaut.http.client` and see how it is redirecting?

**Reply by rafael-coelho-amaro** [reply]
> Can you add a logger for `io.micronaut.http.client` and see how it is redirecting?

hi, the io.micronaut.http.client log was:
```
13:35:27.130 [main] INFO  io.micronaut.runtime.Micronaut - Startup completed in 3424ms. Server Running: http://localhost:8080
13:39:40.174 [default-nioEventLoopGroup-1-3] TRACE i.m.h.server.netty.NettyHttpServer - Server localhost:8080 Received Request: GET /hybris/v2/users/anonymous/carts/16d4ceb2?fields=FULL
13:39:40.179 [default-nioEventLoopGroup-1-3] DEBUG i.m.h.s.netty.RoutingInBoundHandler - Request GET /hybris/v2/users/anonymous/carts/16d4ceb2?fields=FULL
13:39:40.184 [default-nioEventLoopGroup-1-3] DEBUG i.m.h.s.netty.RoutingInBoundHandler - No matching route: GET /hybris/v2/users/anonymous/carts/16d4ceb2?fields=FULL
13:39:41.134 [default-nioEventLoopGroup-1-4] DEBUG i.m.h.client.netty.DefaultHttpClient - Sending HTTP GET to https://localhost:8083/hybris/v2/users/anonymous/carts/16d8cda4-9363-45bb-9909-f2cb734ac66b?fields=FULL
13:39:41.136 [default-nioEventLoopGroup-1-4] TRACE i.m.h.client.netty.DefaultHttpClient - Authorization: Bearer xDAndZzj6gNp0bxcrFMOt37k8co
13:39:41.137 [default-nioEventLoopGroup-1-4] TRACE i.m.h.client.netty.DefaultHttpClient - User-Agent: PostmanRuntime/7.26.8
13:39:41.137 [default-nioEventLoopGroup-1-4] TRACE i.m.h.client.netty.DefaultHttpClient - Accept: */*
13:39:41.137 [default-nioEventLoopGroup-1-4] TRACE i.m.h.client.netty.DefaultHttpClient - Postman-Token: f71xe0b4-fs22-8131-998a-fb5fc20xa2vb
13:39:41.137 [default-nioEventLoopGroup-1-4] TRACE i.m.h.client.netty.DefaultHttpClient - Host: localhost:8080
13:39:41.137 [default-nioEventLoopGroup-1-4] TRACE i.m.h.client.netty.DefaultHttpClient - Accept-Encoding: gzip, deflate, br
13:39:41.137 [default-nioEventLoopGroup-1-4] TRACE i.m.h.client.netty.DefaultHttpClient - connection: close
13:39:41.380 [default-nioEventLoopGroup-1-4] ERROR i.m.h.s.netty.RoutingInBoundHandler - Unexpected error occurred: Invalid HTTP status code: 421
```

I put the code in github https://github.com/rafael-coelho-amaro/proxy-app-sample

**Comment by rafael-coelho-amaro** [comment]
What I need to know is how that proxy works? Does it really redirect the message or makes a new call with the same header and body? With it makes a new call, then I cannot use as a proxy, if it really redirect the message, why the SSL is broken?

Also, Am I using that "proxy" right? If it not, there are somewhere I can get a better sample and better description on how does that works?

**Reply by graemerocher** [reply]
Maybe create an issue with a link to your sample and steps to reproduce. Thanks
